source/extensions/http/cache/file_system_http_cache/DESIGN.md

```diff
@@ -0,0 +1,58 @@
+* Cache entry files are named `cache-` followed by a stable hash key for the entry.
+<a name="tree-structure"></a>
+* (When implemented) the tree structure of folders is simply one level deep of folders named `cache-0000`, `cache-0001` etc. as four-digit hexadecimal numbers up to the configured number of subdirectories. Cache files are placed in a folder according to a short stable hash of their key. On cache startup, any cache entries found to be in the wrong folder (as would be the case if the number of folders was reconfigured) will simply be removed.
+
+## Discussions
+
+### Startup
+
+In the current implementation, on startup all cache entries are enumerated synchronously to provide an initially-accurate measure of the cache size. If server startup time is found to be impacted too significantly by this, the operation could be made asynchronous and moved to the start of the eviction thread. It is still useful to do some minimal operations on startup, so as to be able to detect and reject misconfigurations.
+
+<a name="thundering-herd"></a>
+### Thundering herd
+
+The current implementation, if there are multiple requests for the same resource before the cache is populated, has only one of them perform an insert operation to the cache, and the rest simply bypass the cache. This can cause the "thundering herd" problem - if requests come in bursts the cache will not protect the upstream from that load.
+
+One possible solution would be to have all requesters for the same cache entry stream as the cache entry is written. However, if we do that, and the inserting stream gets closed prematurely, all the dependent streams would be forced to drop their also-incomplete responses.
+
+Another possible solution is to have secondary requesters wait until the cache entry is populated or abandoned before deciding whether to become cache readers or inserters (or bypass, if it turns out to be uncacheable). The downside of this option is that for large content, the dependent clients won't start streaming at all until the first client *finishes* streaming.
+
+An ideal solution would be to either make an inserting stream non-cancellable (i.e. if the client cancels, the upstream connection continues to stream to populate the cache). This could be achieved either by using the existing stream and adding a "non-cancellable" feature in the core (a bit of a large scale change), or by making insertion not use the existing stream at all, instead creating its own client. The problem with that option is that ideally the client would only pass through the filters that are upstream of the cache filter, and there is currently no mechanism for creating a new "partial filter chain" client like this.
+
+_Proposal from jmarantz:_
+
+The lock object could go into one of several states:
```

Comment:
First, I strongly support the current approach of initially ignoring the problem. The thundering herd issue is complicated, with lots of reasonable approaches and tradeoffs to be considered. There's plenty of value in getting a filesystem-based cache into production, and we shouldn't delay it for this.

Some observations:
1. We must first decide whether to pause a request. The proposed list of states is a good start, and I like the approach of letting the config set this policy, because request coalescing isn't always desirable.
2. We should support optionally starting a paused response before the cache fill is complete, to avoid unreasonably delaying large/streaming responses.
3. The cache fill from upstream should either be gated by the fastest client, or should not be gated (except by how fast the cache can insert). Otherwise, a slow first client will slow down all clients.
4. Upstream connection problems are tricky.
  a. When should we restart an upstream connection?
  b. When should we report errors to clients?
  c. When should we let waiting requests just proceed upstream?
  d. If we resume an upstream request, how do we confirm that it's equal and what do we do if it isn't?
  e. Should we run multiple requests in parallel in case one fails?
5. How do we handle conditional requests?
6. How do we handle range requests?
7. If we support pausing requests before the first request has response headers, how do we handle Vary?
8. All this should be managed by CacheFilter, with minimal requirements placed on HttpCache implementations. However, it might make sense to initially implement it in a particular HttpCache implementation, then move it to CacheFilter once it's settled and tested. That would make it easier to evolve, at the cost of later work to move it.

Some of these are tricky; others are fairly straightforward, but they all involve tradeoffs. Many of them must be decided before any real work can begin on the feature.

Reply:
Also, a major form of the thundering herd problem (the cache validation thundering herd) can be solved much more simply with support for asynchronously validating soon-to-expire entries, or with stale-while-revalidate. That may be enough for many deployments.